File: DSA/Linkedlist/Code/InsertionInLinkedList.cs
// Insertion in Linked List in C#

using System;

class Node {
    public int data;
    public Node next;
    
    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedListInsertion {
    Node head;
    
    void InsertAtPosition(int data, int pos) {
        if (pos < 1) {
            Console.WriteLine("Invalid position");
            return;
        }
        
        Node newNode = new Node(data);
        
        if (pos == 1) {
            newNode.next = head;
            head = newNode;
            return;
        }
        
        Node temp = head;
        for (int i = 1; i < pos - 1 && temp != null; i++)
            temp = temp.next;
        
        if (temp == null) {
            Console.WriteLine("Position out of range");
            return;
        }
        
        newNode.next = temp.next;
        temp.next = newNode;
    }
    
    void InsertSorted(int data) {
        Node newNode = new Node(data);
        
        if (head == null || head.data >= data) {
            newNode.next = head;
            head = newNode;
            return;
        }
        
        Node temp = head;
        while (temp.next != null && temp.next.data < data)
            temp = temp.next;
        
        newNode.next = temp.next;
        temp.next = newNode;
    }
    
    void Display() {
        Node temp = head;
        while (temp != null) {
            Console.Write(temp.data + " -> ");
            temp = temp.next;
        }
        Console.WriteLine("NULL");
    }
    
    static void Main() {
        LinkedListInsertion list = new LinkedListInsertion();
        
        Console.WriteLine("=== Insertion in Linked List ===\n");
        
        // Test 1: Insert at position
        list.InsertAtPosition(10, 1);
        list.InsertAtPosition(20, 2);
        list.InsertAtPosition(30, 3);
        list.InsertAtPosition(15, 2);
        
        Console.Write("After position-based insertions: ");
        list.Display();
        
        LinkedListInsertion list2 = new LinkedListInsertion();
        
        // Test 2: Insert in sorted order
        list2.InsertSorted(20);
        list2.InsertSorted(10);
        list2.InsertSorted(30);
        list2.InsertSorted(15);
        list2.InsertSorted(25);
        
        Console.Write("After sorted insertions: ");
        list2.Display();
        
        Console.WriteLine("\nComplexity Analysis:");
        Console.WriteLine("Insert at position: O(n)");
        Console.WriteLine("Insert sorted: O(n)");
        Console.WriteLine("Space: O(1) for operation, O(n) for list");
    }
}

File: DSA/Linkedlist/Code/InsertionInLinkedList.c
#include <stdio.h>
#include <stdlib.h>

// Insertion in Linked List in C
typedef struct Node {
    int data;
    struct Node* next;
} Node;

Node* createNode(int data) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->data = data;
    node->next = NULL;
    return node;
}

void insertAtPosition(Node** head, int data, int pos) {
    if (pos < 1) {
        printf("Invalid position\n");
        return;
    }
    
    Node* newNode = createNode(data);
    
    if (pos == 1) {
        newNode->next = *head;
        *head = newNode;
        return;
    }
    
    Node* temp = *head;
    for (int i = 1; i < pos - 1 && temp != NULL; i++)
        temp = temp->next;
    
    if (temp == NULL) {
        printf("Position out of range\n");
        free(newNode);
        return;
    }
    
    newNode->next = temp->next;
    temp->next = newNode;
}

// Insert in sorted order
void insertSorted(Node** head, int data) {
    Node* newNode = createNode(data);
    
    if (*head == NULL || (*head)->data >= data) {
        newNode->next = *head;
        *head = newNode;
        return;
    }
    
    Node* temp = *head;
    while (temp->next != NULL && temp->next->data < data)
        temp = temp->next;
    
    newNode->next = temp->next;
    temp->next = newNode;
}

void display(Node* head) {
    Node* temp = head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

void freeList(Node* head) {
    Node* temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }
}

int main() {
    Node* head = NULL;
    
    printf("=== Insertion in Linked List ===\n\n");
    
    // Test 1: Insert at position
    insertAtPosition(&head, 10, 1);
    insertAtPosition(&head, 20, 2);
    insertAtPosition(&head, 30, 3);
    insertAtPosition(&head, 15, 2);
    
    printf("After position-based insertions: ");
    display(head);
    
    freeList(head);
    head = NULL;
    
    // Test 2: Insert in sorted order
    insertSorted(&head, 20);
    insertSorted(&head, 10);
    insertSorted(&head, 30);
    insertSorted(&head, 15);
    insertSorted(&head, 25);
    
    printf("After sorted insertions: ");
    display(head);
    
    printf("\nComplexity Analysis:\n");
    printf("Insert at position: O(n)\n");
    printf("Insert sorted: O(n)\n");
    printf("Space: O(1) for operation, O(n) for list\n");
    
    freeList(head);
    return 0;
}
